File: musiclibrary/src/main/java/com/lzx/musiclibrary/playback/player/MediaPlayback.java
package com.lzx.musiclibrary.playback.player;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.media.AudioManager;
import android.media.MediaPlayer;
import android.media.PlaybackParams;
import android.os.Build;
import android.text.TextUtils;

import com.danikula.videocache.HttpProxyCacheServer;
import com.lzx.musiclibrary.MusicService;
import com.lzx.musiclibrary.aidl.model.SongInfo;
import com.lzx.musiclibrary.cache.CacheConfig;
import com.lzx.musiclibrary.cache.CacheUtils;
import com.lzx.musiclibrary.constans.State;
import com.lzx.musiclibrary.manager.FocusAndLockManager;
import com.lzx.musiclibrary.utils.BaseUtil;
import com.lzx.musiclibrary.utils.LogUtil;
import com.lzx.musiclibrary.utils.SPUtils;

import java.io.IOException;

import static com.lzx.musiclibrary.constans.Constans.play_back_pitch;
import static com.lzx.musiclibrary.constans.Constans.play_back_speed;
import static com.lzx.musiclibrary.manager.FocusAndLockManager.AUDIO_NO_FOCUS_CAN_DUCK;
import static com.lzx.musiclibrary.manager.FocusAndLockManager.AUDIO_NO_FOCUS_NO_DUCK;
import static com.lzx.musiclibrary.manager.FocusAndLockManager.VOLUME_DUCK;
import static com.lzx.musiclibrary.manager.FocusAndLockManager.VOLUME_NORMAL;

/**
 * Created by xian on 2018/1/20.
 */

public class MediaPlayback implements Playback,
        FocusAndLockManager.AudioFocusChangeListener,
        MediaPlayer.OnPreparedListener,
        MediaPlayer.OnCompletionListener,
        MediaPlayer.OnErrorListener, MediaPlayer.OnBufferingUpdateListener, MediaPlayer.OnSeekCompleteListener {

    private boolean isOpenCacheWhenPlaying = false;
    private boolean mPlayOnFocusGain;
    private boolean mAudioNoisyReceiverRegistered;
    private boolean mExoPlayerNullIsStopped = false;
    private String mCurrentMediaId; //当前播放的媒体id
    private long currbufferedPosition = 0;
    private SongInfo mCurrentMediaSongInfo;
    private boolean isGiveUpAudioFocusManager = false;
    private MediaPlayer mMediaPlayer;
    private Callback mCallback;
    private Context mContext;
    private FocusAndLockManager mFocusAndLockManager;

    private int mPlayState = State.STATE_IDLE;

    private HttpProxyCacheServer mProxyCacheServer;
    private HttpProxyCacheServer.Builder builder;


    public MediaPlayback(Context context, CacheConfig cacheConfig, boolean isGiveUpAudioFocusManager) {
        Context applicationContext = context.getApplicationContext();
        this.mContext = applicationContext;
        mFocusAndLockManager = new FocusAndLockManager(applicationContext, this);
        this.isGiveUpAudioFocusManager = isGiveUpAudioFocusManager;
        builder = CacheUtils.createHttpProxyCacheServerBuilder(mContext, cacheConfig);
        if (cacheConfig != null && cacheConfig.isOpenCacheWhenPlaying()) {
            isOpenCacheWhenPlaying = true;
        }
        mProxyCacheServer = builder.build();
    }

    private final IntentFilter mAudioNoisyIntentFilter = new IntentFilter(AudioManager.ACTION_AUDIO_BECOMING_NOISY);

    private final BroadcastReceiver mAudioNoisyReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            if (AudioManager.ACTION_AUDIO_BECOMING_NOISY.equals(intent.getAction())) {
                if (isPlaying()) {
                    Intent i = new Intent(context, MusicService.class);
                    mContext.startService(i);
                }
            }
        }
    };

    private void releaseResources(boolean releasePlayer) {
        // Stops and releases player (if requested and available).
        if (releasePlayer && mMediaPlayer != null) {
            mMediaPlayer.stop();
            mMediaPlayer.reset();
            mMediaPlayer.release();
            mMediaPlayer = null;
            mExoPlayerNullIsStopped = true;
            mPlayOnFocusGain = false;
        }
        mFocusAndLockManager.releaseWifiLock();
    }

    private void registerAudioNoisyReceiver() {
        if (!mAudioNoisyReceiverRegistered) {
            mContext.registerReceiver(mAudioNoisyReceiver, mAudioNoisyIntentFilter);
            mAudioNoisyReceiverRegistered = true;
        }
    }

    private void unregisterAudioNoisyReceiver() {
        if (mAudioNoisyReceiverRegistered) {
            mContext.unregisterReceiver(mAudioNoisyReceiver);
            mAudioNoisyReceiverRegistered = false;
        }
    }

    private void configurePlayerState() {
        if (mFocusAndLockManager.getCurrentAudioFocusState() == AUDIO_NO_FOCUS_NO_DUCK) {
            if (!isGiveUpAudioFocusManager) {
                pause();
            }
        } else {
            registerAudioNoisyReceiver();
            if (mFocusAndLockManager.getCurrentAudioFocusState() == AUDIO_NO_FOCUS_CAN_DUCK) {
                mMediaPlayer.setVolume(VOLUME_DUCK, VOLUME_DUCK);
            } else {
                mMediaPlayer.setVolume(VOLUME_NORMAL, VOLUME_NORMAL);
            }
            if (mPlayOnFocusGain) {
                mMediaPlayer.start();
                mPlayState = State.STATE_PLAYING;
                mPlayOnFocusGain = false;
                if (mCallback != null) {
                    mCallback.onPlaybackStatusChanged(mPlayState);
                }
            }
        }
    }

    @Override
    public void start() {

    }

    @Override
    public void stop(boolean notifyListeners) {
        mFocusAndLockManager.giveUpAudioFocus();
        unregisterAudioNoisyReceiver();
        releaseResources(true);
        mPlayState = State.STATE_STOP;
    }

    @Override
    public void setState(int state) {

    }

    @Override
    public int getState() {
        int state;
        if (mMediaPlayer == null) {
            state = State.STATE_IDLE;
        } else {
            state = mPlayState;
        }
        return state;
    }

    @Override
    public boolean isConnected() {
        return true;
    }

    @Override
    public boolean isPlaying() {
        return mPlayOnFocusGain || (mMediaPlayer != null && mPlayState == State.STATE_PLAYING);
    }

    @Override
    public long getCurrentStreamPosition() {
        return mMediaPlayer != null ? mMediaPlayer.getCurrentPosition() : 0;
    }

    @Override
    public long getBufferedPosition() {
        return currbufferedPosition;
    }

    @Override
    public void updateLastKnownStreamPosition() {

    }

    @Override
    public void play(SongInfo info) {
        mPlayOnFocusGain = true;
        mFocusAndLockManager.tryToGetAudioFocus();
        registerAudioNoisyReceiver();
        String mediaId = info.getSongId();
        boolean mediaHasChanged = !TextUtils.equals(mediaId, mCurrentMediaId);
        if (mediaHasChanged) {
            mCurrentMediaId = mediaId;
            mCurrentMediaSongInfo = info;
        }

        if (mediaHasChanged || mMediaPlayer == null) {
            releaseResources(false); // release everything except the player

            String source = info.getSongUrl();
            if (source != null && BaseUtil.isOnLineSource(source)) {
                source = source.replaceAll(" ", "%20"); // Escape spaces for URLs
            }

            String playUrl;
            if (BaseUtil.isOnLineSource(source)) {
                if (isOpenCacheWhenPlaying) {
                    playUrl = mProxyCacheServer.getProxyUrl(source);
                } else {
                    playUrl = source;
                }
            } else {
                playUrl = source;
            }
            if (TextUtils.isEmpty(playUrl)) {
                if (mCallback != null) {
                    mCallback.onError("song url is null");
                }
                return;
            }

            //LogUtil.i("isOpenCacheWhenPlaying = " + isOpenCacheWhenPlaying + " playUri = " + playUrl);

            if (mMediaPlayer == null) {
                mMediaPlayer = new MediaPlayer();
                //当装载流媒体完毕的时候回调
                mMediaPlayer.setOnPreparedListener(this);
                //当流媒体播放完毕的时候回调
                mMediaPlayer.setOnCompletionListener(this);
                //当播放中发生错误的时候回调
                mMediaPlayer.setOnErrorListener(this);
                mMediaPlayer.setOnBufferingUpdateListener(this);
                mMediaPlayer.setOnSeekCompleteListener(this);

                changePlaybackParameters();
            }

            try {
                mMediaPlayer.reset();
                mMediaPlayer.setDataSource(playUrl);
                mMediaPlayer.prepareAsync();
                mPlayState = State.STATE_ASYNC_LOADING;
                if (mCallback != null) {
                    mCallback.onPlaybackStatusChanged(mPlayState);
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
            mFocusAndLockManager.acquireWifiLock();
        } else {
            configurePlayerState();
        }
    }

    private void changePlaybackParameters() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            float spSpeed = (float) SPUtils.get(mContext, play_back_speed, 1f);
            float spPitch = (float) SPUtils.get(mContext, play_back_pitch, 1f);
            float currSpeed = mMediaPlayer.getPlaybackParams().getSpeed();
            float currPitch = mMediaPlayer.getPlaybackParams().getPitch();
            if (spSpeed != currSpeed || spPitch != currPitch) {
                setPlaybackParameters(spSpeed, spPitch);
            }
        }
    }

    @Override
    public void onPrepared(MediaPlayer mp) {
        configurePlayerState();
    }

    @Override
    public void pause() {
        if (mMediaPlayer != null) {
            mMediaPlayer.pause();
            mPlayState = State.STATE_PAUSED;
            if (mCallback != null) {
                mCallback.onPlaybackStatusChanged(mPlayState);
            }
        }
        // While paused, retain the player instance, but give up audio focus.
        releaseResources(false);
        unregisterAudioNoisyReceiver();
    }

    @Override
    public void seekTo(long position) {
        if (mMediaPlayer != null) {
            registerAudioNoisyReceiver();
            mMediaPlayer.seekTo((int) position);
        }
    }

    @Override
    public void setCurrentMediaId(String mediaId) {
        mCurrentMediaId = mediaId;
    }

    @Override
    public String getCurrentMediaId() {
        return mCurrentMediaId;
    }

    @Override
    public int getDuration() {
        return mMediaPlayer != null ? mMediaPlayer.getDuration() : 0;
    }

    @Override
    public SongInfo getCurrentMediaSongInfo() {
        return mCurrentMediaSongInfo;
    }

    @Override
    public void openCacheWhenPlaying(boolean isOpen) {
        isOpenCacheWhenPlaying = isOpen;
    }

    @Override
    public void setPlaybackParameters(float speed, float pitch) {
        if (mMediaPlayer != null && Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            PlaybackParams params = new PlaybackParams();
            params.setSpeed(speed);
            params.setPitch(pitch);
            mMediaPlayer.setPlaybackParams(params);
        }
    }

    @Override
    public void setVolume(float audioVolume) {
        if (mMediaPlayer != null) {
            mMediaPlayer.setVolume(audioVolume, audioVolume);
        }
    }

    @Override
    public int getAudioSessionId() {
        if (mMediaPlayer != null) {
            return mMediaPlayer.getAudioSessionId();
        }
        return 0;
    }

    @Override
    public float getPlaybackSpeed() {
        if (mMediaPlayer != null && Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            return mMediaPlayer.getPlaybackParams().getSpeed();
        } else {
            return (float) SPUtils.get(mContext, play_back_speed, 1f);
        }
    }

    @Override
    public float getPlaybackPitch() {
        if (mMediaPlayer != null && Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            return mMediaPlayer.getPlaybackParams().getPitch();
        } else {
            return (float) SPUtils.get(mContext, play_back_pitch, 1f);
        }
    }


    @Override
    public void setCallback(Callback callback) {
        this.mCallback = callback;
    }

    @Override
    public void onAudioFocusLossTransient() {
        mPlayOnFocusGain = mMediaPlayer != null && mPlayState == State.STATE_PLAYING;
    }

    @Override
    public void onAudioFocusChange() {
        if (mMediaPlayer != null) {
            // Update the player state based on the change
            configurePlayerState();
        }
    }

    @Override
    public void onCompletion(MediaPlayer mp) {
        if (mCallback != null) {
            mCallback.onPlayCompletion();
            //mCallback.onPlaybackStatusChanged(State.STATE_ENDED);
        }
    }

    @Override
    public boolean onError(MediaPlayer mp, int what, int extra) {
        if (mCallback != null) {
            mCallback.onError("MediaPlayer error " + what);
        }
        return false;
    }

    @Override
    public void onBufferingUpdate(MediaPlayer mp, int percent) {
        if (isOpenCacheWhenPlaying && mCurrentMediaSongInfo != null) {
            boolean fullyCached = mProxyCacheServer.isCached(mCurrentMediaSongInfo.getSongUrl());
            currbufferedPosition = fullyCached ? getDuration() : percent * getDuration();
        } else {
            currbufferedPosition = percent * getDuration();
        }
    }

    @Override
    public void onSeekComplete(MediaPlayer mediaPlayer) {
       //TODO
        LogUtil.i("-------------------onSeekComplete------------------");
    }
}

File: app/src/main/java/com/lzx/musiclib/example/MediaPlayback.kt
package com.lzx.musiclib.example

import android.content.Context
import android.media.MediaPlayer
import com.lzx.musiclib.example.FocusAndLockManager.Companion.AUDIO_NO_FOCUS_CAN_DUCK
import com.lzx.musiclib.example.FocusAndLockManager.Companion.AUDIO_NO_FOCUS_NO_DUCK
import com.lzx.musiclib.example.FocusAndLockManager.Companion.VOLUME_DUCK
import com.lzx.musiclib.example.FocusAndLockManager.Companion.VOLUME_NORMAL
import com.lzx.starrysky.playback.offline.ICache
import com.lzx.starrysky.playback.player.Playback
import com.lzx.starrysky.provider.SongInfo
import com.lzx.starrysky.utils.StarrySkyUtils
import java.io.IOException

/**
 * MediaPlayer 简单demo
 */
class MediaPlayback(context: Context, private val playbackCache: ICache?) : Playback,
    MediaPlayer.OnPreparedListener,
    MediaPlayer.OnCompletionListener,
    MediaPlayer.OnErrorListener,
    MediaPlayer.OnBufferingUpdateListener,
    FocusAndLockManager.AudioFocusChangeListener {

    private var mediaPlayer: MediaPlayer? = null
    private var mCallback: Playback.Callback? = null
    private var mPlayState: Int = Playback.STATE_IDLE

    private var mMediaPlayerNullIsStopped = false
    private var mPlayOnFocusGain = false
    private var currBufferedPosition: Long = 0
    private var mCurrentMediaId: String = ""
    private var mCurrVolume: Float = 0F
    private var mErrorProgress: Long = 0
    private var currSongInfo: SongInfo? = null
    private var mFocusAndLockManager: FocusAndLockManager? = null

    init {
        mFocusAndLockManager = FocusAndLockManager(context, this)
    }

    override val playbackState: Int
        get() = if (mediaPlayer == null) Playback.STATE_IDLE else mPlayState

    override val isConnected: Boolean
        get() = true

    override val isPlaying: Boolean
        get() = mPlayOnFocusGain || (mediaPlayer != null && mPlayState == Playback.STATE_PLAYING)

    override val currentStreamPosition: Long
        get() = mediaPlayer?.currentPosition?.toLong() ?: 0L

    override val bufferedPosition: Long
        get() = currBufferedPosition

    override val duration: Long
        get() = mediaPlayer?.duration?.toLong() ?: 0

    override var currentMediaId: String
        get() = mCurrentMediaId
        set(value) {
            mCurrentMediaId = value
        }

    override var volume: Float
        get() = mCurrVolume
        set(value) {
            mCurrVolume = value
            mediaPlayer?.setVolume(value, value)
        }

    override val currPlayInfo: SongInfo?
        get() = currSongInfo

    override val audioSessionId: Int
        get() = mediaPlayer?.audioSessionId ?: 0

    private fun releaseResources(releasePlayer: Boolean) {
        // Stops and releases player (if requested and available).
        if (releasePlayer) {
            mediaPlayer?.stop()
            mediaPlayer?.reset()
            mediaPlayer?.release()
            mediaPlayer = null
            mMediaPlayerNullIsStopped = true
            mPlayOnFocusGain = false
        }
        mFocusAndLockManager?.releaseWifiLock()
    }

    override fun stop() {
        mFocusAndLockManager?.giveUpAudioFocus()
        releaseResources(true)
        mPlayState = Playback.STATE_STOPPED
        mCallback?.onPlaybackStatusChanged(currSongInfo, mPlayState)
    }

    override fun play(songInfo: SongInfo, isPlayWhenReady: Boolean) {
        mPlayOnFocusGain = true
        mFocusAndLockManager?.tryToGetAudioFocus()

        val mediaId = songInfo.songId
        if (mediaId.isEmpty()) {
            return
        }
        currSongInfo = songInfo
        val mediaHasChanged = mediaId != currentMediaId
        if (mediaHasChanged) {
            currentMediaId = mediaId
        }
        if (mediaHasChanged || mediaPlayer == null) {
            //url 处理
            var source = songInfo.songUrl
            if (source.isEmpty()) {
                mCallback?.onPlaybackError(currPlayInfo, "播放 url 为空")
                return
            }
            source = source.replace(" ".toRegex(), "%20") // Escape spaces for URL
            //代理url
            val proxyUrl = playbackCache?.getProxyUrl(source)
            source = if (proxyUrl.isNullOrEmpty()) source else proxyUrl
            playbackCache?.startCache(source)

            if (mediaPlayer == null) {
                mediaPlayer = MediaPlayer()
                //当装载流媒体完毕的时候回调
                mediaPlayer?.setOnPreparedListener(this)
                //当流媒体播放完毕的时候回调
                mediaPlayer?.setOnCompletionListener(this)
                //当播放中发生错误的时候回调
                mediaPlayer?.setOnErrorListener(this)
                mediaPlayer?.setOnBufferingUpdateListener(this)
//            changePlaybackParameters()
            }
            try {
                mediaPlayer?.reset()
                mediaPlayer?.setDataSource(source)
                mediaPlayer?.prepareAsync()
                mPlayState = Playback.STATE_BUFFERING
                mCallback?.onPlaybackStatusChanged(currSongInfo, mPlayState)
            } catch (e: IOException) {
                e.printStackTrace()
            }
            mFocusAndLockManager?.acquireWifiLock()
        } else {
            configurePlayerState()
        }
    }

    override fun onPrepared(mp: MediaPlayer?) {
        configurePlayerState()
    }

    override fun onCompletion(mp: MediaPlayer?) {
        mPlayState = Playback.STATE_IDLE
        mCallback?.onPlaybackCompletion()
    }

    override fun onError(mp: MediaPlayer?, what: Int, extra: Int): Boolean {
        StarrySkyUtils.log("onError = $what")
        mCallback?.onPlaybackError(currSongInfo, "MediaPlayer error $what")
        mPlayState = Playback.STATE_ERROR
        mErrorProgress = currBufferedPosition
        mCurrentMediaId = ""
        return false
    }

    override fun onBufferingUpdate(mp: MediaPlayer?, percent: Int) {
        currBufferedPosition = percent * duration
    }

    override fun pause() {
        mediaPlayer?.pause()
        mPlayState = Playback.STATE_PAUSED
        mCallback?.onPlaybackStatusChanged(currSongInfo, mPlayState)
        releaseResources(false)
    }

    override fun seekTo(position: Long) {
        mediaPlayer?.seekTo(position.toInt())
    }

    override fun onFastForward() {
        //
    }

    override fun onRewind() {
        //
    }

    override fun onDerailleur(refer: Boolean, multiple: Float) {
        //
    }

    override fun setCallback(callback: Playback.Callback) {
        mCallback = callback
    }

    override fun onAudioFocusLossTransient() {
        mPlayOnFocusGain = mediaPlayer != null && mPlayState == Playback.STATE_PLAYING
    }

    override fun onAudioFocusChange() {
        mediaPlayer?.let {
            configurePlayerState()
        }
    }

    private fun configurePlayerState() {
        if (mFocusAndLockManager?.currentAudioFocusState == AUDIO_NO_FOCUS_NO_DUCK) {
            pause()
        } else {
            if (mFocusAndLockManager?.currentAudioFocusState == AUDIO_NO_FOCUS_CAN_DUCK) {
                mediaPlayer?.setVolume(VOLUME_DUCK, VOLUME_DUCK)
            } else {
                mediaPlayer?.setVolume(VOLUME_NORMAL, VOLUME_NORMAL)
            }
            if (mPlayOnFocusGain) {
                mediaPlayer?.start()
                mPlayState = Playback.STATE_PLAYING
                mPlayOnFocusGain = false
                mCallback?.onPlaybackStatusChanged(currSongInfo, mPlayState)
                if (mErrorProgress != 0L) {
                    seekTo(mErrorProgress)
                    mErrorProgress = 0
                }
            }
        }
    }
}